---
title: "Creation of expr_list object"
author: "Manczinger M."
date: "8/9/2021"
output: html_document
---

This script creates the expr_list_tcem object, which contains the gene expression of each gene encoding each TCEM (see the PNAS publication for details)

Please note that the created expr_list object may differ from the one uploaded to Mendeley, because the data source of the GenomicFeatures and org.Hs.eg.db library could differ from the one used in the study (the original object was created on 31st January 2020)

```{r}
library(edgeR)
library(GenomicFeatures)
library(org.Hs.eg.db)
library(protr)
library(fastmatch)
```

#1. preprocessing

```{r}
expression = read.delim("GSE127209_human_TEC_unadj.counts.txt", sep = "\t", stringsAsFactors = F) # gene expression data of cTECs were generated by the study with DOI: 10.1172/jci.insight.125377. The file can be downloaded from Mendeley Data (dx.doi.org/10.17632/63xr979845.1).
expression = expression[,grepl("cTEC", colnames(expression))]
expression_normalized = calcNormFactors(expression)
for(i in 1 : ncol(expression)) expression[,i] = round(expression[,i] / expression_normalized[i])
hg19.ens = makeTxDbFromUCSC(genome="hg19", tablename="ensGene")
exonic = exonsBy(hg19.ens, by="gene")
red.exonic = reduce(exonic)
exon.lengths = sum(width(red.exonic))
rpkm = rpkm(expression, gene.length = exon.lengths[rownames(expression)])
genexpr = apply(rpkm, 1, median)
```

#2. gene id conversion

```{r}
swiss_to_ensg = read.delim("uniprot_to_hugo.tab", stringsAsFactors = F) # The file was downloaded from the uniprot database on 31st January 2020. The file can be downloaded from Mendeley Data (dx.doi.org/10.17632/63xr979845.1).
proteome = readFASTA("uniprot-proteome_UP000005640+reviewed_yes.fasta") # The file was downloaded from the uniprot database on 27th January 2020. The file can be downloaded from Mendeley Data (dx.doi.org/10.17632/63xr979845.1).

swiss_to_ensg = swiss_to_ensg[sapply(proteome, nchar) >=9,]
proteome = proteome[sapply(proteome, nchar) >=9]
names(proteome) = sapply(names(proteome), FUN = function(x) unlist(strsplit(x, "\\|"))[3])
names(proteome) = gsub("_HUMAN", "", names(proteome))

names = mapIds(org.Hs.eg.db, names(genexpr), 'SYMBOL', 'ENSEMBL')
genexpr = cbind(genexpr, symbol = names[names(genexpr)])
genexpr = aggregate(as.numeric(genexpr[,1]) ~ genexpr[,2], FUN = function(x) max(x, na.rm = T))
table(is.na(match(swiss_to_ensg$Gene.names...primary.., names)))
genexpr_for_proteome = genexpr[,2]
names(genexpr_for_proteome) = genexpr[,1]
genexpr_for_proteome = genexpr_for_proteome[unique(swiss_to_ensg$Gene.names...primary..)]
names(genexpr_for_proteome) = unique(swiss_to_ensg$Gene.names...primary..)

#genexpr_for_proteome #This file is also on github.
```

#3. matching gene expression with TCEMs

```{r}
proteome = readFASTA("uniprot-proteome_UP000005640+reviewed_yes.fasta") #The file can be downloaded from Mendeley Data (dx.doi.org/10.17632/63xr979845.1).
proteome = proteome[sapply(proteome, nchar) >=9]

proteome_sub9 = sapply(proteome, FUN = function(x) substring(x, 1:(nchar(x) - 4), c(9:nchar(x), rep(nchar(x), 4))))
proteome_sub9_tcem = proteome_sub9[sapply(proteome_sub9, length) > 4]
for(i in 1 : length(proteome_sub9_tcem)) proteome_sub9_tcem[[i]] = proteome_sub9_tcem[[i]][-1:-3]
for(i in 1 : length(proteome_sub9_tcem)) proteome_sub9_tcem[[i]] = proteome_sub9_tcem[[i]][-length(proteome_sub9_tcem[[i]])]
for(i in 1 : length(proteome_sub9_tcem)) {
  proteome_sub9_tcem[[i]] = cbind(proteome_sub9_tcem[[i]], genexpr_for_proteome[swiss_to_ensg$Gene.names...primary..[i]])
  proteome_sub9_tcem[[i]][,1] = substr(proteome_sub9_tcem[[i]][,1], 1, 5)
}
proteome_sub9_tcem = do.call(rbind, proteome_sub9_tcem)
tcem_expression = proteome_sub9_tcem
```

#4. arrange TCEM expression to list

```{r}
elements_expr = aggregate(as.numeric(tcem_expression[,2]) ~ tcem_expression[,1], FUN = function(x) x) #It is a long-time running

variations = expand.grid(rownames(AABLOSUM62), rownames(AABLOSUM62), rownames(AABLOSUM62), rownames(AABLOSUM62), rownames(AABLOSUM62))
variations = apply(variations, 1, FUN = function(x) paste0(x, collapse = ""))

expr_list = as.list(rep(NA,20^5))
names(expr_list) = variations

elements_expr = elements_expr[!grepl("U", elements_expr[,1]),]
expr_list[elements_expr[,1]] = sapply(elements_expr[,2], unlist)
```

