---
title: "Dataset 1 and 2 creation"
author: "Balazs Koncz"
date: '01 09 2021'
output: html_document
---
#Setup

```{r include=FALSE}
#setwd("D:/ImObsRecognition/") #set your working directory
library(dplyr)
library(magrittr)
library(pbapply)
library(fastmatch)
library(data.table)
library(protr)
library(stringi)
library(Rfast)
```

#Dataset 1 assembly
##T cell activation assays

```{r}
tcell_full = fread("tcell_full_v3.csv", skip = 1) #the file can be downloaded from Mendeley Data (dx.doi.org/10.17632/63xr979845.1)
tcell_full = tcell_full[,c(125,12,1,11,106)] #selecting of relevant columns
colnames(tcell_full) = c("allele", "peptide", "assay", "object_type", "qualitative_measure")

tcell_filt = tcell_full %>% 
  filter(nchar(peptide) == 9 | nchar(peptide) == 10) %>%
  filter(object_type == "Linear peptide") %>%
  filter(substr(allele,1,5) == "HLA-A" | substr(allele,1,5) == "HLA-B" | substr(allele,1,5) == "HLA-C") %>%
  filter(nchar(allele) == 11) %>%
  dplyr::select(-object_type) %>% 
  transform(allele = gsub("HLA-|:|\\*", "", allele))
tcell_filt = tcell_filt[sapply(tcell_filt$peptide, function(x) protcheck(x)),] #no. of peptides after filtering step 1 (on FigS1): 48,651
rm(tcell_full)

#Aggregating data for individual allele-peptide pairs
activating = tcell_filt %>% 
  mutate(qm = ifelse(qualitative_measure == "Negative", "neg", "pos")) %>% 
  dplyr::select(-qualitative_measure) %>% 
  group_by(allele, peptide, qm) %>% 
  dplyr::summarise(n = n()) %>% 
  tidyr::spread(key = qm, value = n) %>% 
  mutate_at(vars(neg, pos), function(x) x = ifelse(is.na(x), 0, x)) %>% 
  mutate(n_activating_assay = neg+pos) %>% 
  mutate(activating_ratio = pos/n_activating_assay) %>%
  ungroup() %>% 
  dplyr::select(-c("neg", "pos")) #no. of peptides after filtering step 2: 30,443
rm(tcell_filt)
```

##Step 3 left branch in FigS1
Predicting HLA-binding with NetMHCpan-4.0.
Selecting pairs with either < 500 nM binding affinity or < 2% binding rank percentile

```{r eval=FALSE, include=FALSE}
load("binding_matrix") #the file can be downloaded from Mendeley Data (dx.doi.org/10.17632/63xr979845.1)
rownames(binding_matrix$aff) = gsub("HLA-|:", "", rownames(binding_matrix$aff))
rownames(binding_matrix$rp) = gsub("HLA-|:", "", rownames(binding_matrix$rp))
activating$aff = apply(activating,1,function(x) ifelse(x[1] %in% rownames(binding_matrix$aff), binding_matrix$aff[x[1],x[2]], NA))
activating$bindingrp = apply(activating,1,function(x) ifelse(x[1] %in% rownames(binding_matrix$rp), binding_matrix$rp[x[1],x[2]], NA))
#HLA-B44:01 prediction is missing
rm(binding_matrix)
ligands_raw_d1 = activating %>% 
  filter(aff<500|bindingrp<2) %>% 
  dplyr::select(allele, peptide, n_activating_assay, activating_ratio) #no. of peptides after filtering step 3: 21,474
```

##Step 4 in FigS1
Keeping pairs, if T cell activation was tested in at least two assays

```{r}
ligands_raw_d1 %<>% 
  filter(n_activating_assay >= 2) #no. of peptides after filtering step 4: 5,630
```

##Step 5 in FigS1
Keeping pairs, if the peptide was not found in the reference human proteome

```{r}
proteome = readFASTA("uniprot-proteome_UP000005640+reviewed_yes.fasta") #the file can be downloaded from Mendeley Data (dx.doi.org/10.17632/63xr979845.1)
tempdf = ligands_raw_d1 %>% dplyr::select(peptide) %>% unique()
tempdf$human = pbsapply(tempdf$peptide, function(z) any(stri_detect_fixed(proteome, z))) #TRUE - human, FALSE - nonhuman
table(tempdf$human)
humaneps = tempdf$peptide[tempdf$human == T]
ligands_raw_d1 %<>% 
  filter(!peptide %in% humaneps) #no. of peptides after filtering step 5: 4,634
rm(proteome, tempdf, humaneps)
```

##Step 6 in FigS1
Defining immunogenic and nonimmunogenic pairs and excluding the ones with controversial assay results

```{r}
ligands_raw_d1$immunogenicity = NA
ligands_raw_d1$immunogenicity[ligands_raw_d1$activating_ratio == 0] = 0
ligands_raw_d1$immunogenicity[ligands_raw_d1$activating_ratio > .5] = 1
ligands_raw_d1 %<>% na.omit()
table(ligands_raw_d1$immunogenicity)

#Controversial assay results
length(unique(ligands_raw_d1$peptide[duplicated(ligands_raw_d1$peptide)])) #95
#ligands_raw_d1 %>% group_by(peptide) %>% dplyr::summarise(n = n(), m = mean(immunogenicity)) %>% View()
drop_peps = ligands_raw_d1 %>% dplyr::group_by(peptide) %>% dplyr::summarise(m = mean(immunogenicity)) %>% filter(m > 0, m < 1) %>% pull(peptide) #13 peptides/95
ligands_raw_d1 %<>% filter(!peptide %in% drop_peps) #no. of peptides after filtering step 6: 4,089
rm(drop_peps)
```

##Step 7 in FigS1
Dropping peptides that have conflicting results in pair with different alleles

```{r}
peps = unique(ligands_raw_d1$peptide)
for(j in 1:length(peps)) {
  tempdf = ligands_raw_d1 %>% filter(peptide == peps[j])
  if(nrow(tempdf) == 1) next()
  tempdf = tempdf[which.max(tempdf$n_activating_assay),]
  ligands_raw_d1[ligands_raw_d1$peptide == peps[j],] = tempdf
}
rm(peps, tempdf, j)
ligands_raw_d1 = unique(ligands_raw_d1) #no. of peptides after filtering step 7: 3,955

ligands_raw_d1$immunogenicity = factor(ligands_raw_d1$immunogenicity, levels = c("1","0"))
```

#Dataset 2 assembly
##Binding assays from IEDB

```{r}
mhc_ligand_full = fread("mhc_ligand_full.csv", skip = 1) #the file can be downloaded from Mendeley Data (dx.doi.org/10.17632/63xr979845.1)
mhc_ligand_full = mhc_ligand_full[,c(96,12,1,11,84)] #selecting of relevant columns
colnames(mhc_ligand_full) = c("allele", "peptide", "assay", "object_type", "qualitative_measure")

mhc_ligand_filt = mhc_ligand_full %>%
  filter(nchar(peptide) == 9 | nchar(peptide) == 10) %>%
  filter(object_type == "Linear peptide") %>%
  filter(substr(allele,1,5) == "HLA-A" | substr(allele,1,5) == "HLA-B" | substr(allele,1,5) == "HLA-C") %>%
  filter(nchar(allele) == 11) %>%
  dplyr::select(-object_type) %>% 
  transform(allele = gsub("HLA-|:|\\*", "", allele))
mhc_ligand_filt = mhc_ligand_filt[sapply(mhc_ligand_filt$peptide, function(x) protcheck(x)),] #469,417
rm(mhc_ligand_full)

#Aggregating data for individual allele-peptide pairs
binding = mhc_ligand_filt %>% 
  mutate(qm = ifelse(qualitative_measure == "Negative", "neg", "pos")) %>% 
  dplyr::select(-qualitative_measure) %>% 
  group_by(allele, peptide, qm) %>% 
  dplyr::summarise(n = n()) %>% 
  tidyr::spread(key = qm, value = n) %>% 
  mutate_at(vars(neg, pos), function(x) x = ifelse(is.na(x), 0, x)) %>% 
  mutate(n_binding_assay = neg+pos) %>% 
  mutate(binding_ratio = pos/n_binding_assay) %>%
  ungroup() %>% 
  dplyr::select(-c("neg", "pos")) #320,055

rm(mhc_ligand_filt)
```

##Step 3 right branch in FigS1
Keeping pairs, if:
- HLA-binding of the given peptide was tested in at least two assays and
- fraction of positive binding assay results is higher than 0.6

```{r}
ligands_raw_d2 = binding %>% 
  left_join(activating, by = c("allele", "peptide")) %>% 
  filter(!is.na(activating_ratio))
rm(activating, binding)
ligands_raw_d2 %<>% 
  filter(n_binding_assay >= 2, binding_ratio > 0.6) #no. of peptides after filtering step 3: 2,345
```

##Step 4 in FigS1
Keeping pairs, if T cell activation was tested in at least two assays

```{r}
ligands_raw_d2 %<>% 
  filter(n_activating_assay >= 2) #no. of peptides after filtering step 4: 1,506
```

##Step 5 in FigS1
Keeping pairs, if the peptide was not found in the reference human proteome

```{r}
proteome = readFASTA("uniprot-proteome_UP000005640+reviewed_yes.fasta") #the file can be downloaded from Mendeley Data (dx.doi.org/10.17632/63xr979845.1)
tempdf = ligands_raw_d2 %>% dplyr::select(peptide) %>% unique()
tempdf$human = pbsapply(tempdf$peptide, function(z) any(stri_detect_fixed(proteome, z))) #TRUE - human, FALSE - nonhuman
table(tempdf$human)
humaneps = tempdf$peptide[tempdf$human == T]
ligands_raw_d2 %<>% filter(!peptide %in% humaneps) #no. of peptides after filtering step 5: 1,318
rm(proteome, tempdf, humaneps)
```

##Step 6 in FigS1
Defining immunogenic and nonimmunogenic pairs and excluding the ones with controversial assay results

```{r}
ligands_raw_d2$immunogenicity = NA
ligands_raw_d2$immunogenicity[ligands_raw_d2$activating_ratio == 0] = 0
ligands_raw_d2$immunogenicity[ligands_raw_d2$activating_ratio > .5] = 1
ligands_raw_d2 = ligands_raw_d2[!is.na(ligands_raw_d2$immunogenicity),]
table(ligands_raw_d2$immunogenicity)

#Controversial assay results
length(unique(ligands_raw_d2$peptide[duplicated(ligands_raw_d2$peptide)])) #19
#ligands_raw_d2 %>% group_by(peptide) %>% dplyr::summarise(n = n(), m = mean(immunogenicity)) %>% View()
drop_peps = ligands_raw_d2 %>% dplyr::group_by(peptide) %>% dplyr::summarise(m = mean(immunogenicity)) %>% filter(m > 0, m < 1) %>% pull(peptide) #1 peptide(YAQMWSLMY) / 19
ligands_raw_d2 %<>% filter(!peptide %in% drop_peps) #no. of peptides after filtering step 6: 1,115
rm(drop_peps)
```

##Step 7 in FigS1
Dropping peptides that have conflicting results in pair with different alleles

```{r}
peps = unique(ligands_raw_d2$peptide)
for(j in 1:length(peps)) {
  tempdf = ligands_raw_d2 %>% filter(peptide == peps[j])
  if(nrow(tempdf) == 1) next()
  tempdf = tempdf[which.max(tempdf$n_activating_assay),]
  ligands_raw_d2[ligands_raw_d2$peptide == peps[j],] = tempdf
}
rm(peps, tempdf, j)
ligands_raw_d2 = unique(ligands_raw_d2) #no. of peptides after filtering step7: 1,087

ligands_raw_d2$immunogenicity = factor(ligands_raw_d2$immunogenicity, levels = c("1","0"))
```

#Step 8 right branch in FigS1
Excluding similar peptides using an iterative method
You have to download and install clustal omega from http://www.clustal.org/omega/#Download

```{r}
clomega <- function(epitopes, similarity) {
  strings = as.vector(rbind(paste0(">", epitopes), epitopes))
  writeLines(strings, "epitopes.fasta")
  system("\"C:\\Users\\User\\Downloads\\clustal-omega-1.2.2-win64\\clustalo.exe\" -i epitopes.fasta -o out_epitopes.fasta -v --distmat-out=dist.txt --threads=8 --full --force") #set the location of clustal omega
  distmtx = try(readLines("dist.txt"))
  maximum = max(nchar(distmtx))
  distmtx = subset(distmtx, nchar(distmtx) == maximum)
  start = gregexpr(" ", distmtx[1])
  start = start[[1]][length(start[[1]])-length(distmtx) + 1] + 1
  distmtx = substr(distmtx, start, maximum)
  distmtx = strsplit(distmtx, " ")
  distmtx = matrix(unlist(distmtx), nrow = length(distmtx[[1]]), ncol = length(distmtx[[1]]))
  class(distmtx) = "numeric"
  diag(distmtx) = NA
  minimum = min(distmtx, na.rm = TRUE)
  rownames(distmtx) = epitopes
  colnames(distmtx) = epitopes
  while (min(distmtx, na.rm = TRUE) < similarity) {
    print(paste(min(distmtx, na.rm = TRUE), length(epitopes)))
    wm = which(distmtx == min(distmtx, na.rm = TRUE), arr.ind = TRUE)
    means = colMeans(distmtx, na.rm = TRUE)
    means = means[wm[,1]]
    index = wm[which.min(means),1]
    distmtx = distmtx[-index, -index]
    epitopes = epitopes[-index]
  }
  return(distmtx)
}

dist_matr = clomega(epitopes = ligands_raw_d2$peptide, similarity = 0.5)
ligands_raw_d2 %>% filter(!peptide %in% rownames(dist_matr)) %>% pull(peptide) -> dropped_eps_clo_d2 #452 peptides
#writeLines(dropped_eps_clo_d2, "dropped_eps_clo_d2.txt")
rm(dropped_eps_clo_d2)

ligands_nonhuman_d2 = ligands_raw_d2 %<>% 
  filter(peptide %in% rownames(dist_matr)) #no. of peptides after filtering step 8: 635
rm(dist_matr, clomega)
table(ligands_raw_d2$immunogenicity) #1: 360, 0: 275
```

#Step 8 left branch in FigS1
Drop peptides of dataset 2

```{r}
ligands_nonhuman_d1 = ligands_raw_d1[!ligands_raw_d1$peptide %in% ligands_nonhuman_d2$peptide,] #no. of peptides after filtering step 8: 3,380
table(ligands_nonhuman_d1$immunogenicity) #1 - 1,093; 0 - 2,287
rm(ligands_raw_d1, ligands_raw_d2)
```

#Pentamer frequency
Output file (pentamerfreq_tcem) can be downloaded from Mendeley Data (dx.doi.org/10.17632/63xr979845.1)

```{r}
proteome = readFASTA("uniprot-proteome_UP000005640+reviewed_yes.fasta") #the file can be downloaded from Mendeley Data (dx.doi.org/10.17632/63xr979845.1)
nonamers = sapply(proteome, function(x) substring(x, 1:(nchar(x)-8), 9:nchar(x)))
nonamers = unlist(nonamers, use.names = F)
nonamers = nonamers[nchar(nonamers) == 9]
tcems = substr(nonamers,4,8)
tcems = tcems[!grepl("U", tcems)]

pfreqs = Table(tcems)
pentamers = tidyr::unite(data = expand.grid(rownames(AABLOSUM62),rownames(AABLOSUM62),rownames(AABLOSUM62),rownames(AABLOSUM62),rownames(AABLOSUM62)), col = "pentamer", sep = "")
pentamers = pentamers$pentamer

zerofreqs = rep(0, length(pentamers)-length(pfreqs))
names(zerofreqs) = setdiff(pentamers, names(pfreqs))

pentamerfreq_tcem = c(pfreqs, zerofreqs)
pentamerfreq_tcem = pentamerfreq_tcem[pentamers]
rm(proteome, nonamers, pentamers, pfreqs, tcems, zerofreqs)
```

#Sequence similarity
You have to download and install blast from https://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/
output file (bl62score) can be downloaded from github
```{r}
source("sequence_similarity_func.R") #R file can be downloaded from github
create_db_proteome(blast_folder = "C:/NCBI/blast-2.11.0+/bin/", 
                   proteome_fasta_loc = "uniprot-proteome_UP000005640+reviewed_yes.fasta", 
                   peptide_length = 9, 
                   db_name = "human_proteome_9mers")
blastp(blast_folder = "C:/NCBI/blast-2.11.0+/bin/", 
       query_peptides = c(dataset_1$epitope[nchar(dataset_1$epitope)==9],dataset_2$epitope[nchar(dataset_2$epitope)==9]), 
       db_loc = "C:/NCBI/blast-2.11.0+/bin/human_proteome_9mers.fasta", 
       out_loc = "blastpout_nonamers", 
       threads = 6)
bl62score_nonamers = calcSeqSim(blastpout_loc = "blastpout_nonamers")


create_db_proteome(blast_folder = "C:/NCBI/blast-2.11.0+/bin/", 
                   proteome_fasta_loc = "uniprot-proteome_UP000005640+reviewed_yes.fasta", 
                   peptide_length = 9, 
                   db_name = "human_proteome_10mers")
blastp(blast_folder = "C:/NCBI/blast-2.11.0+/bin/", 
       query_peptides = c(dataset_1$epitope[nchar(dataset_1$epitope)==10],dataset_2$epitope[nchar(dataset_2$epitope)==10]), 
       db_loc = "C:/NCBI/blast-2.11.0+/bin/human_proteome_10mers.fasta", 
       out_loc = "blastpout_dekamers", 
       threads = 6)
bl62score_dekamers = calcSeqSim(blastpout_loc = "blastpout_dekamers")

bl62score = c(bl62score_nonamers, bl62score_dekamers)
```

#Add other variables
output files (ligands_nonhuman_d1, ligands_nonhuman_d2, ligands_nonhuman_d1d2) can be downloaded from github

```{r}
##TCEM frequency
ligands_nonhuman_d1$tcem = unname(pbsapply(ligands_nonhuman_d1$peptide, function(x) ifelse(nchar(x) == 9, substr(x,4,8), substr(x,5,9))))
load("pentamerfreq_tcem") #the file can be downloaded from Mendeley Data (dx.doi.org/10.17632/63xr979845.1)
ligands_nonhuman_d1$freqpenta_tcem = pentamerfreq_tcem[ligands_nonhuman_d1$tcem]
rm(pentamerfreq_tcem)

##Expression
load("expr_list_tcem") #the file can be downloaded from Mendeley Data (dx.doi.org/10.17632/63xr979845.1)
ligands_nonhuman_d1$pentamer_expression = unname(pbsapply(ligands_nonhuman_d1$tcem, function(x) median(expr_list[[fmatch(x, names(expr_list))]])))
rm(expr_list)

##Çleavage scores
load("tpr_score") #the file can be downloaded from Mendeley Data (dx.doi.org/10.17632/63xr979845.1)
ligands_nonhuman_d1$thymopscore = unname(pbsapply(ligands_nonhuman_d1$tcem, function(x) median(score_list[[fmatch(x, names(score_list))]])))

load("ipr_score") #the file can be downloaded from Mendeley Data (dx.doi.org/10.17632/63xr979845.1)
ligands_nonhuman_d1$immunopscore = unname(pbsapply(ligands_nonhuman_d1$tcem, function(x) median(score_list[[fmatch(x, names(score_list))]])))
rm(score_list)

#Blosum62 similarity score
load("bl62score") #the file can be downloaded from github
ligands_nonhuman_d1$bl62score = unname(bl62score[match(ligands_nonhuman_d1$peptide, names(bl62score))])
rm(bl62score)

#Housekeeping ratio
load("tcem_housekeeping_stat") #the file can be downloaded from Mendeley Data (dx.doi.org/10.17632/63xr979845.1)
elements_housekeeping %<>% 
  as.data.frame(stringsAsFactors = F) %>% 
  mutate_at(2:3, as.numeric)
ligands_nonhuman_d1$housekeeping_ratio = unname(pbsapply(ligands_nonhuman_d1$tcem, function(x) ifelse(x %in% elements_housekeeping$motif, elements_housekeeping$housekeeping[elements_housekeeping$motif == x] / (elements_housekeeping$housekeeping[elements_housekeeping$motif == x]+elements_housekeeping$`not housekeeping`[elements_housekeeping$motif == x]),NA))) #It is a long-time running
rm(elements_housekeeping)

ligands_nonhuman_d1 %<>% 
  select(allele, peptide, immunogenicity, tcem, freqpenta_tcem, pentamer_expression, thymopscore, immunopscore, bl62score, housekeeping_ratio)

#Dataset 2
##TCEM frequency
ligands_nonhuman_d2$tcem = unname(pbsapply(ligands_nonhuman_d2$peptide, function(x) ifelse(nchar(x) == 9, substr(x,4,8), substr(x,5,9))))
load("pentamerfreq_tcem") #the file can be downloaded from Mendeley Data (dx.doi.org/10.17632/63xr979845.1)
ligands_nonhuman_d2$freqpenta_tcem = pentamerfreq_tcem[ligands_nonhuman_d2$tcem]
rm(pentamerfreq_tcem)

##Expression
load("expr_list_tcem") #the file can be downloaded from Mendeley Data (dx.doi.org/10.17632/63xr979845.1)
ligands_nonhuman_d2$pentamer_expression = unname(pbsapply(ligands_nonhuman_d2$tcem, function(x) median(expr_list[[fmatch(x, names(expr_list))]])))
rm(expr_list)

##Çleavage scores
load("tpr_score") #the file can be downloaded from Mendeley Data (dx.doi.org/10.17632/63xr979845.1)
ligands_nonhuman_d2$thymopscore = unname(pbsapply(ligands_nonhuman_d2$tcem, function(x) median(score_list[[fmatch(x, names(score_list))]])))

load("ipr_score") #the file can be downloaded from Mendeley Data (dx.doi.org/10.17632/63xr979845.1)
ligands_nonhuman_d2$immunopscore = unname(pbsapply(ligands_nonhuman_d2$tcem, function(x) median(score_list[[fmatch(x, names(score_list))]])))
rm(score_list)

#Blosum62 similarity score
load("bl62score") #the file can be downloaded from github
ligands_nonhuman_d2$bl62score = unname(bl62score[match(ligands_nonhuman_d2$peptide, names(bl62score))])
rm(bl62score)

#Housekeeping ratio
load("tcem_housekeeping_stat") #the file can be downloaded from Mendeley Data (dx.doi.org/10.17632/63xr979845.1)
elements_housekeeping %<>% 
  as.data.frame(stringsAsFactors = F) %>% 
  mutate_at(2:3, as.numeric)
ligands_nonhuman_d2$housekeeping_ratio = unname(pbsapply(ligands_nonhuman_d2$tcem, function(x) ifelse(x %in% elements_housekeeping$motif, elements_housekeeping$housekeeping[elements_housekeeping$motif == x] / (elements_housekeeping$housekeeping[elements_housekeeping$motif == x]+elements_housekeeping$`not housekeeping`[elements_housekeeping$motif == x]),NA))) #It is a long-time running
rm(elements_housekeeping)

ligands_nonhuman_d2 %<>% 
  select(allele, peptide, immunogenicity, tcem, freqpenta_tcem, pentamer_expression, thymopscore, immunopscore, bl62score, housekeeping_ratio)

#Datasets 1 & 2
load("ligands_nonhuman_d1") #the file can be downloaded from github
ligands_nonhuman_d1 = cbind(ligands_nonhuman_d1, dataset = "d1")
load("ligands_nonhuman_d2") #the file can be downloaded from github
ligands_nonhuman_d2 = cbind(ligands_nonhuman_d2, dataset = "d2")
ligands_nonhuman = rbind(ligands_nonhuman_d1, ligands_nonhuman_d2)
rm(ligands_nonhuman_d1, ligands_nonhuman_d2)
ligands_nonhuman %<>% 
    mutate(group = case_when(
    .$dataset == "d1" & .$immunogenicity == 0 ~ "d1_nim",
    .$dataset == "d1" & .$immunogenicity == 1 ~ "d1_im",
    .$dataset == "d2" & .$immunogenicity == 0 ~ "d2_nim",
    .$dataset == "d2" & .$immunogenicity == 1 ~ "d2_im",
  )) %>% 
  transform(group = factor(group, levels = c("d1_im", "d1_nim", "d2_im", "d2_nim")))

```

