---
title: "Generating files for pathogen analysis"
author: "Manczinger M."
date: "9/13/2021"
output: html_document
---

The script generates objects that are needed for generating Figure 5. 

```{r}
library(fastmatch)

species = list.files("pathogen_9mers/") #list pathogen files containing proteins decomposed to 9-mers. The file can be downloaded from Mendeley Data (dx.doi.org/10.17632/63xr979845.1)
load("tpr_score") #The file can be downloaded from Mendeley Data (dx.doi.org/10.17632/63xr979845.1) or generated by the calculation_of_thymoproteasomal_and_immunoproteasomal_cleavage_scores.Rmd script
thymomedians = sapply(score_list, FUN = function(x) median(x, na.rm = T))
load("pentamerfreq_tcem") #the file can be downloaded from Mendeley Data (dx.doi.org/10.17632/63xr979845.1) or generated by the data_construction.Rmd script
load("expr_list_tcem") #the file can be downloaded from Mendeley Data (dx.doi.org/10.17632/63xr979845.1) or generated by the creation_of_expr_list_object.Rmd script
exprmedians = sapply(expr_list, FUN = function(x) median(x, na.rm = T))
load("uniprot_id_to_species_name")
load("cutoffs") # the file contains the median number 9mer peptides bound by all HLA-A or B or C alleles from the proteome of different pathogens

files = list.files("uniprot_proteomes_hla_binding/") #the total size of the files is > 10 GB, we will share them upon request
files_used = files[sapply(files, FUN = function(x) any(unlist(strsplit(x, "\\."))[1] == species))]

exprcutoff = 0.203075 #33th percentile of peptide TCEM expression in the naive CD8+ T cell analysis of healthy individuals
thymocutoff = 0.914470 #25th percentile of peptide TCEM thymoproteasome score in the naive CD8+ T cell analysis of healthy individuals

res = lapply(species, FUN = function(x) {
  load(paste0("pathogen_9mers/", x))
  tcems = substr(eps, 4, 8)
  tcemfreq = pentamerfreq_tcem[fmatch(tcems, names(pentamerfreq_tcem))]
  tcemexpr = exprmedians[fmatch(tcems, names(exprmedians))]
  tcemthymo = thymomedians[fmatch(tcems, names(thymomedians))]
  rare = tcemfreq < 4
  lowexpr = tcemexpr < exprcutoff
  lowexpr[is.na(lowexpr)] = F
  lowthymo = tcemthymo < thymocutoff
  lowthymo[is.na(lowthymo)] = F
  isna = !is.na(rare)
  files_sub = files[grepl(x, files)]
  mtx = t(sapply(files_sub, FUN = function(y) {
    load(paste0("uniprot_proteomes_hla_binding/", y)) #the total size of the files is > 10 GB, we will share them upon request
    affrank = order(results$aff)
    rankrank = order(results$rp)
    strong_aff = results$aff < 50 & isna
    strong_rp = results$rp < 0.5 & isna
    strong_aff_strong_rp = strong_aff & strong_rp
    if(!any(strong_aff_strong_rp)) strong_aff_strong_rp[affrank <= cutoffs[paste0(x, "_", substr(unlist(strsplit(y, "\\."))[2], 1, 1))]] = T
    if(!any(strong_aff_strong_rp)) strong_aff_strong_rp[affrank <= 1] = T
    tempvector = c(sum(strong_aff_strong_rp), sum((rare | lowexpr | lowthymo) & strong_aff_strong_rp) / sum(strong_aff_strong_rp))
  }))
})
res = do.call(rbind, res)
res = as.data.frame(res, stringsAsFactors = F)
colnames(res) = c("num_aff_strong_rp_strong","aff_strong_rp_strong")

res$species = sapply(rownames(res), FUN = function(x) unlist(strsplit(x, "\\."))[1])
res$alleles = sapply(rownames(res), FUN = function(x) unlist(strsplit(x, "\\."))[2])
res$species_name = uniprot_id_to_species_name[fmatch(res$species, uniprot_id_to_species_name[,1]),2]

save(res, file = "np_tcems_hla_pathogens")

load("np_tcems_hla_pathogens")

cwd = read.delim("cwd200_alleles.txt", stringsAsFactors = F, skip = 1) # the file can be downloaded from http://igdawg.org/pubs/cwd200.zip
cwd = cwd[substr(cwd$IMGT.HLA.3.9.0.Allele.Name, 1, 2) == "A*" | substr(cwd$IMGT.HLA.3.9.0.Allele.Name, 1, 2) == "B*" | substr(cwd$IMGT.HLA.3.9.0.Allele.Name, 1, 2) == "C*",]
alleles = cwd$IMGT.HLA.3.9.0.Allele.Name[cwd$CWD.2.0.0.Category == "C"]
alleles = substr(alleles, 1, 7)
alleles = gsub("\\*|:", "", alleles)

hepatitis_b = res[res$species_name == "Hepatitis B virus",]
hepatitis_b = hepatitis_b[!is.na(match(hepatitis_b$alleles, alleles)),]
hepatitis_c = res[res$species_name == "Hepatitis C virus genotype 1a",]
hepatitis_c = hepatitis_c[!is.na(match(hepatitis_c$alleles, alleles)),]
hpv = res[res$species_name == "Human papillomavirus type 16",]
hpv = hpv[!is.na(match(hpv$alleles, alleles)),]
dengue = res[res$species_name == "Dengue virus type 1",]
dengue = dengue[!is.na(match(dengue$alleles, alleles)),]

summary = rbind(cbind("HPV", "B57", "susc", hpv$aff_strong_rp_strong[substr(hpv$alleles, 1, 3) == "B57"], hpv$alleles[substr(hpv$alleles, 1, 3) == "B57"]),
                cbind("HPV", "B47", "susc", hpv$aff_strong_rp_strong[substr(hpv$alleles, 1, 3) == "B47"], hpv$alleles[substr(hpv$alleles, 1, 3) == "B47"]),
                cbind("HPV", "C04", "prot", hpv$aff_strong_rp_strong[substr(hpv$alleles, 1, 3) == "C04"], hpv$alleles[substr(hpv$alleles, 1, 3) == "C04"]),
                cbind("HBV", "B07", "prot", hepatitis_b$aff_strong_rp_strong[substr(hepatitis_b$alleles, 1, 3) == "B07"], hepatitis_b$alleles[substr(hepatitis_b$alleles, 1, 3) == "B07"]),
                cbind("HBV", "B58", "prot", hepatitis_b$aff_strong_rp_strong[substr(hepatitis_b$alleles, 1, 3) == "B58"], hepatitis_b$alleles[substr(hepatitis_b$alleles, 1, 3) == "B58"]),
                cbind("HBV", "A31", "prot", hepatitis_b$aff_strong_rp_strong[substr(hepatitis_b$alleles, 1, 3) == "A31"], hepatitis_b$alleles[substr(hepatitis_b$alleles, 1, 3) == "A31"]), #optional
                cbind("HBV", "B13", "susc", hepatitis_b$aff_strong_rp_strong[substr(hepatitis_b$alleles, 1, 3) == "B13"], hepatitis_b$alleles[substr(hepatitis_b$alleles, 1, 3) == "B13"]), #optional
                cbind("HCV", "B40", "prot", hepatitis_c$aff_strong_rp_strong[substr(hepatitis_c$alleles, 1, 3) == "B40"], hepatitis_c$alleles[substr(hepatitis_c$alleles, 1, 3) == "B40"]),
                cbind("HCV", "B61", "prot", hepatitis_c$aff_strong_rp_strong[hepatitis_c$alleles == "B4002" | hepatitis_c$alleles == "B4003" | hepatitis_c$alleles == "B4004" | hepatitis_c$alleles == "B4006"],
                      hepatitis_c$alleles[hepatitis_c$alleles == "B4002" | hepatitis_c$alleles == "B4003" | hepatitis_c$alleles == "B4004" | hepatitis_c$alleles == "B4006"]),
                cbind("HCV", "C05", "susc", hepatitis_c$aff_strong_rp_strong[substr(hepatitis_c$alleles, 1, 3) == "C05"], hepatitis_c$alleles[substr(hepatitis_c$alleles, 1, 3) == "C05"]),
                cbind("HCV", "C04", "susc", hepatitis_c$aff_strong_rp_strong[substr(hepatitis_c$alleles, 1, 3) == "C04"], hepatitis_c$alleles[substr(hepatitis_c$alleles, 1, 3) == "C04"]),
                cbind("Dengue", "A24", "susc", dengue$aff_strong_rp_strong[substr(dengue$alleles, 1, 3) == "A24"], dengue$alleles[substr(dengue$alleles, 1, 3) == "A24"]),
                cbind("Dengue", "B44", "prot", dengue$aff_strong_rp_strong[substr(dengue$alleles, 1, 3) == "B44"], dengue$alleles[substr(dengue$alleles, 1, 3) == "B44"]))

groups = aggregate(as.numeric(summary[,4]) ~ paste(summary[,1], summary[,2], summary[,3]), FUN = function(x) mean(x, na.rm = T), na.action = na.pass)
groups = cbind(matrix(unlist(strsplit(groups[,1], " ")), ncol = 3, byrow = T), groups[,2])

save(groups, file = "infection_risk_HLA")

```

